File: Data Structures & Algorithms/max-water-container/submission-0.cpp
class Solution {
public:
    int maxArea(vector<int>& heights) {
        
        int left = 0;
        int right = heights.size()-1;
        int maxArea = 0;
        int area;

        while(left < right)
        {
            area = min(heights[left], heights[right]) * (right-left);
            maxArea = max(maxArea, area);
            
            if(heights[left] <= heights[right])
            {
                left++;
            }
            else
            {
                right--;
            }
        }
        
        return maxArea;
    }
};

File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-0.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        stack = []
        maxArea = 0

        for i, h in enumerate(heights):
            start = i
            while stack and h < stack[-1][1]:
                leftIdx, height = stack.pop()
                maxArea = max(maxArea, height * (i - leftIdx))
                start = leftIdx
            stack.append((start, h))
        
        for i, h in stack:
            maxArea = max(maxArea, h * (len(heights) - i))
        return maxArea

